digdag-core/src/main/java/io/digdag/core/workflow/AttemptBuilder.java

```diff
@@ -110,14 +111,26 @@ public AttemptRequest buildFromStoredWorkflow(
         return monitors.build();
     }
 
-    private Config buildSessionParameters(Config overrideParams, Optional<Scheduler> sr, Instant sessionTime, ZoneId timeZone)
+    private Config buildSessionParameters(Config overrideParams, Optional<Scheduler> sr, Instant sessionTime, ZoneId timeZone, Optional<Instant> lastExecutedSessionTime)
     {
         Config params = overrideParams.deepCopy();
         if (sr.isPresent()) {
             Instant lastSessionTime = sr.get().lastScheduleTime(sessionTime).getTime();
             Instant nextSessionTime = sr.get().nextScheduleTime(sessionTime).getTime();
             params.set("last_session_time", formatSessionTime(lastSessionTime, timeZone));
             params.set("next_session_time", formatSessionTime(nextSessionTime, timeZone));
+            if (lastExecutedSessionTime.isPresent()) {
+                // TODO: Come up with a better names for these parameters?
+
+                // This is the last session that was processed. Normally this will be the immediately preceding session,
+                // but it can be an earlier session if some were skipped due to skip_on_overtime = true.
+                params.set("last_processed_session_time", formatSessionTime(lastExecutedSessionTime.get(), timeZone));
+
+                // This is the first session time slot that should be included if the user wants to process any preceding skipped sessions.
+                // Normally this is the same as the current session_time but can be an earlier session if some were skipped.
+                Instant firstUnprocessedSessionTime = sr.get().nextScheduleTime(lastExecutedSessionTime.get()).getTime();
+                params.set("first_unprocessed_session_time", formatSessionTime(firstUnprocessedSessionTime, timeZone));
```

Comment:
I think that we want to have `last_successful_session_time` instead of `last_processed_session_time` or `first_unprocessed_session_time`.
We, Treasure Data, were selling a solution packaged with a BI tool called Metric Insights. When we schedule a query on it, it skips execution on overtime, and it provides a variable similar to the last successful execution time. It was quite useful. I imagine this use case with `skip_on_overtime: true` option.
A typical query was like this:

```sql
select
  TD_TIME_FORMAT(time, 'yyyy-MM-dd', 'JST') AS d,
  count(distinct user_id) as uu
from purchases
where :last_measurement_time < TD_TIME_FORMAT(time, 'yyyy-MM-dd', 'JST')
group by 1
```

`:last_successful_timestamp` is last successfully measured (calculated) timestamp (@toru-takahashi should know about this well). Users run queries every hour or every day depending on computing resource they have.

There're two benefits:

1. People don't have to care much about a single failure because the next execution will recover the results.
2. Backfilling was (relatively) easy because what they needed was to resetting :last_measurement_time to a past timestamp.

2 may not apply to Digdag but 1 is good.

We can get last successful session using this query:

```sql
select sa.*
from sessions s
join session_attempts sa on s.last_attempt_id = sa.id
where project_id = :projectId and workflow_name = :workflowName
and s.session_time < :sessionTime
and bitand(sa.state_flags, `AttemptStateFlags.SUCCESS_CODE`) != 0  -- successful
order by s.session_time desc
limit 1
```